src/execution.jl

```diff
@@ -436,9 +444,8 @@ macro benchmark(args...)
     return esc(
         quote
             local $tmp = $BenchmarkTools.@benchmarkable $(args...)
-            $BenchmarkTools.warmup($tmp)
             $(hasevals(params) ? :() : :($BenchmarkTools.tune!($tmp)))
-            $BenchmarkTools.run($tmp)
+            $BenchmarkTools.run($tmp; warmup=$(hasevals(params)))
```

Comment:
Here IIUC `hasevals` checks whether `evals` has been set. Your reasoning is that if we have applied `tune!` before, we don't need a warmup. But in the situation where `evals` is set from the start, `tune!` does nothing and so warmup never runs.

Reply:
Yes, this confused me too a bit, but the situation is as follows:
If evals is set at macro expansion time, tune is not called, so we need to warmup, set warmup = true
if evals is not set at macro expansion time, tune is indeed called, so we don't need to warmup, set warmup = false

Note that we are interpolating whether params has evals into the quoted expression before tune runs.